optaplanner-core/src/main/java/org/optaplanner/core/impl/solver/manager/DefaultSolverManager.java

```diff
@@ -0,0 +1,207 @@
+                if (onException != null) {
+                    eventHandlerExecutorService.submit(() -> onException.accept(throwable.getCause()));
+                }
+            }
+            return eventHandlerExecutorService.submit(() -> onSolvingEnded.accept(solution_));
+        });
+    }
+
+    @Override
+    public void stopSolver(Object problemId) {
+        logger.debug("Stopping solver of problemId ({}).", problemId);
+        SolverTask<Solution_> solverTask;
+        synchronized (this) {
+            if (!isProblemSubmitted(problemId)) {
+                throw new IllegalArgumentException("Problem (" + problemId + ") was not submitted.");
+            }
+            solverTask = problemIdToSolverTaskMap.get(problemId);
+        }
+        if (solverTask != null) {
+            solverTask.stopSolver();
+        }
+    }
+
+    @Override
+    public boolean isProblemSubmitted(Object problemId) {
+        return problemIdToSolverTaskMap.containsKey(problemId);
+    }
+
+    @Override
+    public Solution_ getBestSolution(Object problemId) {
+        logger.debug("Getting best solution of problemId ({}).", problemId);
+        SolverTask<Solution_> solverTask = problemIdToSolverTaskMap.get(problemId);
+        if (solverTask == null) {
+            logger.error("Problem ({}) was not submitted.", problemId);
+            return null;
+        }
+        return solverTask.getBestSolution();
+    }
+
+    @Override
+    public Score<?> getBestScore(Object problemId) {
+        logger.debug("Getting best score of problemId ({}).", problemId);
+        SolverTask<Solution_> solverTask = problemIdToSolverTaskMap.get(problemId);
+        if (solverTask == null) {
+            logger.error("Problem ({}) was not submitted.", problemId);
+            return null;
+        }
+        return solverTask.getBestScore();
+    }
+
+    @Override
+    public SolverStatus getSolverStatus(Object problemId) {
+        logger.debug("Getting solver status of problemId ({}).", problemId);
+        SolverTask<Solution_> solverTask = problemIdToSolverTaskMap.get(problemId);
+        if (solverTask == null) {
+            logger.error("Problem ({}) was not submitted.", problemId);
+            return null;
+        }
+        return solverTask.getSolverStatus();
+    }
+
+    @Override
+    public void close() {
+        shutdown();
+    }
+
+    @Override
+    public void shutdown() {
+        logger.info("Shutting down {}.", DefaultSolverManager.class.getName());
+        ThreadUtils.shutdownAwaitOrKill(solverExecutorService, "", "solverExecutorService");
+        ThreadUtils.shutdownAwaitOrKill(eventHandlerExecutorService, "", "eventHandlerExecutorService");
+        stopSolvers(); // without this: shutdownAwaitOrKill appears to only interrupt currently running tasks, those that are queued start solving
+    }
+
+    private void stopSolvers() {
+        for (SolverTask<Solution_> solverTask : problemIdToSolverTaskMap.values()) {
```

Comment:
Personally, I'd prefer `problemIdToSolverTaskMap.values().forEach(solverTask::stopSolver)`. Shorter, don't have to bother with typing.